### PR TITLE
fix: make UI person prop updates persistent

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Problem

UI person property updates were only updating Postgres

## Changes

Make sure Postgres and CH are both updated by using an event to update properties

## How did you test this code?

Wrote tests, tested manually on my local setup

https://github.com/PostHog/posthog/issues/9856

### CODE REVIEW

File: frontend/src/scenes/persons/personsLogic.ts [hunk 1]
```diff
@@ -176,9 +179,20 @@ export const personsLogic = kea<personsLogicType<PersonFilters, PersonLogicProps
                     action = parsedValue !== undefined ? 'updated' : 'removed'
                 }
 
-                actions.setPerson(person) // To update the UI immediately while the request is being processed
-                const response = await api.update(`api/person/${person.id}`, person)
-                actions.setPerson(response)
+                actions.setPerson({ ...person }) // To update the UI immediately while the request is being processed
```
Comment (yakkomajuri) [inline frontend/src/scenes/persons/personsLogic.ts:182]: note that this was broken before and we were only updating the UI after the request completed

File: frontend/src/scenes/persons/personsLogic.ts [hunk 2]
```diff
@@ -176,9 +179,20 @@ export const personsLogic = kea<personsLogicType<PersonFilters, PersonLogicProps
                     action = parsedValue !== undefined ? 'updated' : 'removed'
                 }
 
-                actions.setPerson(person) // To update the UI immediately while the request is being processed
-                const response = await api.update(`api/person/${person.id}`, person)
```
Comment (macobo) [inline frontend/src/scenes/persons/personsLogic.ts:180]: Confused - so the old endpoint still can get person model out of sync? I don't think we've solved the problem then.

File: frontend/src/scenes/persons/personsLogic.ts [hunk 3]
```diff
@@ -176,9 +179,20 @@ export const personsLogic = kea<personsLogicType<PersonFilters, PersonLogicProps
...
-                actions.setPerson(response)
+                actions.setPerson({ ...person }) // To update the UI immediately while the request is being processed
+
+                try {
+                    await api.create(`api/person/update_property`, {
+                        key,
+                        new_value: newValue,
+                        main_distinct_id: person.distinct_ids[0],
```
Comment (macobo) [inline frontend/src/scenes/persons/personsLogic.ts:188]: This is a weird assumption - why not person_id?

File: frontend/src/scenes/persons/personsLogic.ts (hunk 2 above)
Comment (yakkomajuri) [reply frontend/src/scenes/persons/personsLogic.ts:180]: I'm going step by step here - first updating the UI to behave correctly (which is the main way people use this). I'll then yeet whatever code remains once I map out all the implications.

Note I'm on support hero time so am shipping in increments

File: frontend/src/scenes/persons/personsLogic.ts (hunk 3 above)
Comment (yakkomajuri) [reply frontend/src/scenes/persons/personsLogic.ts:188]: Because we need to send an event

Comment (yakkomajuri) [reply frontend/src/scenes/persons/personsLogic.ts:188]: See how the merge people code path works for example

File: frontend/src/scenes/persons/personsLogic.ts (hunk 2 above)
Comment (macobo) [reply frontend/src/scenes/persons/personsLogic.ts:180]: Are you planning on yeeting this in this PR (is this ready for a review)?

If not, then I can't really approve this without e.g. a ticket indicating the problem.

I'm not sold on this approach either - why not change the original endpoint?

File: frontend/src/scenes/persons/personsLogic.ts (hunk 3 above)
Comment (macobo) [reply frontend/src/scenes/persons/personsLogic.ts:188]: Person pages working off of distinct ids is a historical accident waiting for us to clean it up at some point. Could we get the distinct_id by doing a person lookup in the endpoint instead?

File: frontend/src/scenes/persons/personsLogic.ts (hunk 2 above)
Comment (yakkomajuri) [reply frontend/src/scenes/persons/personsLogic.ts:180]: This is ready for a review - it updates the path for updating a person's property via the UI.

The original endpoint processes a full patch to the person, and I need to make sure I'm aware of what depends on it. It is a more general endpoint where people can patch not just properties. 

One issue at a time.

File: frontend/src/scenes/persons/personsLogic.ts (hunk 3 above)
Comment (yakkomajuri) [reply frontend/src/scenes/persons/personsLogic.ts:188]: That's doable, yes

File: frontend/src/scenes/persons/personsLogic.ts (hunk 2 above)
Comment (macobo) [reply frontend/src/scenes/persons/personsLogic.ts:180]: > One issue at a time.

We're promoting our API more and more and this fix doesn't fix the existing API. We should either:
1. Fix this issue properly in this PR
2. Or create a separate issue for that issue 

That's what I was asking for in the original comment - if you're going for (2) please link that ticket. Without this I can't accept this fix as we're introducing needless pain for ourselves in the future.

Comment (yakkomajuri) [reply frontend/src/scenes/persons/personsLogic.ts:180]: https://github.com/PostHog/posthog/issues/9856

Comment (macobo) [reply frontend/src/scenes/persons/personsLogic.ts:180]: Thank you!

File: frontend/src/scenes/persons/personsLogic.ts (hunk 3 above)
Comment (tiina303) [reply frontend/src/scenes/persons/personsLogic.ts:188]: why should we do a lookup, isn't that more work?

Comment (yakkomajuri) [reply frontend/src/scenes/persons/personsLogic.ts:188]: Actually I see what @macobo is talking about in the sense that we used to have the keys for person pages be distinct IDs, which was a problem. However, I believe `personsLogic` will always contain a distinct ids list so I would agree with @tiina303 - no need for an extra lookup

**Comment by macobo** [comment]
Closing for https://github.com/PostHog/posthog/pull/9974

**Pull request closed without merging**